src/Symfony/Component/DependencyInjection/Definition.php

```diff
@@ -64,6 +64,13 @@ public function setFactory($factory)
             $factory = explode('::', $factory, 2);
         }
 
+        if (is_array($factory) && !isset($factory[0])) {
+            if (null === $this->class) {
+                throw new InvalidArgumentException(sprintf('First value of the $factory array should either be a class name, a reference or the definition class should be set. "%s" given.', isset($factory[0]) ? gettype($factory[0]) : 'NULL'));
+            }
+            $factory[0] = $this->class;
```

Comment:
See also https://github.com/symfony/symfony/pull/20264#discussion_r92919282

I think we should resolve properties at once, in the right order. So everythings works as expected.